Undefined error 0 in FILES_SAVE_PRE handler when mutating buffer

Given following `visrc.lua`:

```lua
 -- trim trailing whitespace before saving                                       
 vis.events.subscribe(vis.events.FILE_SAVE_PRE, function(file, _)                
   local lines = file.lines                                                      
   for i=1, #lines do
     local trimmed = lines[i]:match('^(.-)%s+$')                                 
     if trimmed then lines[i] = trimmed end                                      
   end                                                                           
   -- proceed with saving the file...                                            
   return true                                                                   
 end) 
```
I see the info message:

```sh
Can't write `/home/oliver/.config/vis/visrc.lua': Undefined error: 0
```

To reproduce:

1. Put trailing whitespace somewhere in a file.
2. Save with `:w`.
3. Error appears as info. File is trimmed, but not saved.
4. Save again. No error, file is saved.

Might be related to https://github.com/martanne/vis/issues/652.

```sh
oliver:114$ vis -v
vis v0.5 +curses +lua +tre
```

Confirmed the same behaviour on master:

```sh
oliver:142$ ./vis -v
vis  +curses +lua +tre
```

Btw, `make test` fails for me (OpenBSD-current):

```sh
oliver:147$ make test 
[ -e test/Makefile ] || make test-update
Compiling buffer-test binary
/usr/bin/../lib/crt0.o: In function `_start':
(.text+0x52): undefined reference to `main'
cc: error: linker command failed with exit code 1 (use -v to see invocation)
*** Error 1 in test/core (Makefile:23 'buffer-test': @cc -pipe -Os -ffunction-sections -fdata-sections -fPIE -I. -I../.. -DBUFFER_SIZE=4 -DB...)
*** Error 1 in test (Makefile:2 'test')
*** Error 1 in /home/oliver/Workspace/ansic/vis (Makefile:113 'test')
```

I am not sure how to make this report more helpful, but if you have some instructions here, I'd be happy to make separate issue out of this.
I tried `make CC=gcc test` with same result, and tried to _use -v to see invocation_, but I don't know how to specify a custom linker command.

--- a/sam.c
+++ b/sam.c
@@ -1573,6 +1573,10 @@ static bool cmd_substitute(Vis *vis, Win *win, Command *cmd, const char *argv[],
 	return false;
 }
 
+/* cmd_write stores win->file's contents end emits pre/post events.
+ * If the range r covers the whole file, it is updated to account for
+ * potential file's text mutation by a FILE_SAVE_PRE callback.
+ */
 static bool cmd_write(Vis *vis, Win *win, Command *cmd, const char *argv[], Selection *sel, Filerange *r) {
 	if (!win)
 		return false;
@@ -1582,6 +1586,9 @@ static bool cmd_write(Vis *vis, Win *win, Command *cmd, const char *argv[], Sele
 		return false;
 
 	Text *text = file->text;
+	Filerange range_all = text_range_new(0, text_size(text));
+	bool write_entire_file = text_range_equal(r, &range_all);
+
 	const char *filename = argv[1];
 	if (!filename)
 		filename = file->name;
@@ -1599,6 +1606,9 @@ static bool cmd_write(Vis *vis, Win *win, Command *cmd, const char *argv[], Sele
 			vis_info_show(vis, "Rejected write to stdout by pre-save hook");
 			return false;
 		}
+		/* a pre-save hook may have changed the text; need to re-take the range */
+		if (write_entire_file)
+			*r = text_range_new(0, text_size(text));
 
 		bool visual = vis->mode->visual;
 
@@ -1624,8 +1634,7 @@ static bool cmd_write(Vis *vis, Win *win, Command *cmd, const char *argv[], Sele
 			vis_info_show(vis, "WARNING: file will be reduced to active selection");
 			return false;
 		}
-		Filerange all = text_range_new(0, text_size(text));
-		if (!text_range_equal(r, &all)) {
+		if (!write_entire_file) {
 			vis_info_show(vis, "WARNING: file will be reduced to provided range");
 			return false;
 		}
@@ -1648,6 +1657,9 @@ static bool cmd_write(Vis *vis, Win *win, Command *cmd, const char *argv[], Sele
 			vis_info_show(vis, "Rejected write to `%s' by pre-save hook", path);
 			goto err;
 		}
+		/* a pre-save hook may have changed the text; need to re-take the range */
+		if (write_entire_file)
+			*r = text_range_new(0, text_size(text));
 
 		TextSave *ctx = text_save_begin(text, path, file->save_method);
 		if (!ctx) {

